src/internal/models.py

```diff
@@ -314,3 +316,34 @@ class Meta:
 
     def __str__(self):
         return _("“{quote}” —{quoted}").format(quote=self.quote, quoted=self.quoted)
+
+
+def internal_subclass_directory_path(instance: 'Lore', filename: str):
+    model_name = instance._meta.model_name
+    return f"internal/{model_name}s/{filename}"
+
+
+class Lore(models.Model):
+    title = models.CharField(max_length=50, verbose_name=_("title"), null=False)
+    slug = models.SlugField()
```

Comment:
Because this is a wiki, and wikis usually use slugs, not pk's. Yes, it does add complexity to the code, but I still think a slug looks better than a pk in this case.

Reply:
I wouldn't say that other wikis using slugs is really an argument for why we should use them; that being said, the underlying reason is usually either or both of:
1. It prevents enumerating resources, if that's important to the website owners, for whatever reason - usually for security or business reasons
2. It makes it somewhat easy to see at a glance the title of the resource - as long as the URL is not too long or cluttered, of course
   - This is often made redundant, however, as when you paste a URL into e.g. Slack, Facebook or most other social media / messaging services, a small preview of the website is usually shown, which takes info (incl. a webpage title) from the Open Graph data defined on each specific page (see e.g. [how articles and events insert this data](https://github.com/MAKENTNU/web/blob/main/src/news/templates/news/news_base_detail.html#L9-L13))

I guess the latter reason is the one most relevant to us, and it is absolutely a good reason, but I personally think it has more downsides than upsides. The only advantage that I can see is the one mentioned above, but the disadvantages weigh heavier, I think:
1. The main one is that it introduces complexity that adds a maintenance cost
2. As the code is right now, if a user changes the title of a lore article, existing URLs to that article will no longer be valid
   - This can be fixed by making sure that the slug is never changed after a lore article is created, but then you might lose the original advantage of having slugs in the first place, if the title and/or content is changed to no longer be reasonably summarized by the original title/slug

So I'd recommend that you keep the code as simple as it can be (without losing the functionality that you want, of course), but ultimately that decision is for you guys to take 🙂

Reply:
In any case, as the code is now, point 2 above (the bottom one) should be addressed, in my opinion. The alternatives I can think of are:
1. Disallowing changing the title after creation
2. Not updating the slug after changing the title
3. Keeping track of all previous slug values and redirecting from the old ones to the newest one
   * This would also require not allowing any lore articles to be renamed so that their slugs collide with any previous slugs of other articles

Any thoughts on this? 🙂

Reply:
_(Reposted this comment, as the previous one somehow appeared both here and above my most recent review; probably a bug..? 🤷)_

In any case, as the code is now, point 2 above (the bottom one) should be addressed, in my opinion. The alternatives I can think of are:
1. Disallowing changing the title after creation
2. Not updating the slug after changing the title
3. Keeping track of all previous slug values and redirecting from the old ones to the newest one
   * This would also require not allowing any lore articles to be renamed so that their slugs collide with any previous slugs of other articles

Any thoughts on this or additional ideas? 🙂

Reply:
I now created an additional model storing changes in slugs, so URLs consisting of slugs that are not used any more, will still be valid. If two different lore articles at different points in time have had the same slug, the user will be redirected to the one that appear first in the HistoricSlug model.

Reply:
> If two different lore articles at different points in time have had the same slug, the user will be redirected to the one that appear first in the HistoricSlug model.

Hm, this still has the potential of causing confusion when a URL can change which lore article it points to, if a future lore article happens to use the same slug as an existing historical one 🤔 Though it does seems like it would be better than the current solution 🙂

---

Now, you didn't comment on the first part of [my second comment above](#discussion_r1391555887), but I'm assuming that the practical reason you want slugs in the URLs in the first place (apart from the aesthetics of the URLs), is to make it somewhat easy to read the title of the page from the URL - which in turn might act as a way of communicating to users that they're URLs to a kind of wiki (since no parts of the lore URLs contain the word "wiki"). Is this correct?

If so, before going any further with the solution you've already created, I recently had an idea that pretty much eliminates all the problems with using slugs that I've previously mentioned, _and_ should drastically reduce the amount of code! Simply let the slugs be a "decorative" part of the URLs, but have the identifying part of the URLs be the lore articles' PK! This could e.g. look like `https://i.makentnu.no/lore/64/the-title-of-a-lore-article/` (or, alternatively, `https://i.makentnu.no/lore/the-title-of-a-lore-article/64/` - though I personally prefer the first suggestion). Furthermore, lore URLs without the slug or with a wrong/outdated slug can be redirected to the URL with the correct, up-to-date slug. This can be implemented by doing the following:
1. Replace `<slug:slug>` with `<int:pk>` in the `path()` that includes `specific_lore_urlpatterns`
2. Duplicate the `path()` of `LoreDetailView` and add `<str:slug>/` to the path of the new one (so that both paths reference `LoreDetailView`), and placing that path **after** the `change/` and `delete/` paths - since both of those strings would match `<str:slug>/` and therefore essentially make it impossible to reach those two paths
   * Note that I wrote `str:slug`, not `slug:slug`, because we want the view to be able to "correct" the URLs no matter what the last part of the URL is
3. Make sure that the code inside `LoreDetailView` ignores `slug` completely when it queries the database for the lore object, since only the PK will be used to identify lore articles
4. Inside `LoreDetailView`, override its [`get()`](https://docs.djangoproject.com/en/stable/ref/class-based-views/generic-display/#django.views.generic.detail.BaseDetailView.get) method, check the URL's `slug` against the lore object's `slug`, and if they differ, return a redirect response using the correct slug
5. Remove `Lore`'s `slug` field and replace it with a `slug` `@property`, since the slug is not needed anymore for querying, but only for display; also remove overriding `save()` (as mentioned in [this comment](https://github.com/MAKENTNU/web/pull/675/files#r1425896596)), and add `get_absolute_url()` (as mentioned in [this comment](https://github.com/MAKENTNU/web/pull/675/files#r1426077750))
6. This should result in roughly the following code:
   ```python
   # internal/urls.py:

   specific_lore_urlpatterns = [
       path("", views.LoreDetailView.as_view(), name='lore_detail'),  # Redirects to `lore_detail` below
       path("change/", views.LoreUpdateView.as_view(), name='lore_update'),
       path("delete/", views.LoreDeleteView.as_view(), name='lore_delete'),
       # Must be placed at the bottom of this list to not "hide" the paths above
       # (The slug is purely decorative; the PK is the identifying part of the path)
       path("<str:slug>/", views.LoreDetailView.as_view(), name='lore_detail'),
   ]
   lore_urlpatterns = [
       path("", views.LoreListView.as_view(), name='lore_list'),
       path("add/", views.LoreCreateView.as_view(), name='lore_create'),
       path("<int:pk>/", include(specific_lore_urlpatterns)),
   ]

   # internal/views.py:

   class LoreDetailView(DetailView):
       model = Lore
       slug_url_kwarg = None  # Ignore the path's `slug` parameter
       ...  # template_name and context_object_name
   
       def get(self, request, *args, **kwargs):
           if slug := kwargs.get('slug'):
               lore_article = self.get_object()
               if slug != lore_article.slug:
                   return HttpResponseRedirect(lore_article.get_absolute_url())
   
           return super().get(request, *args, **kwargs)
   
       def get_context_data(self, **kwargs):
           ...

   # internal/models.py:

   class Lore(models.Model):
       ...  # fields (with `slug` removed), the `Meta` class, and methods (with `save()` removed)
   
       def get_absolute_url(self):
           return reverse_internal('lore_detail', self.pk, self.slug)
   
       @property
       def slug(self):
           return slugify(self.title, allow_unicode=True)
   ```

What do you think? I haven't tested the code above, so please let me know if you encounter any bugs with it that you're unsure about how to solve :)

If you agree with implementing this idea, I can absolutely understand if you're also hesitant to scrap all the work you've put into your original idea - that's never particularly fun 🙃 Though (and I'm saying this with empathy, as I've been there myself) hopefully this can help remind you in the future to consider discussing the thing you're creating before spending time creating it - for example with the people you plan on inviting to review your work 😊